### PR TITLE
Add Socket Mask for Topology Manager

### PR DESCRIPTION
**Pull request by nolancon**
**What this PR does / why we need it**:
Adding Socket Mask for Topology Manager based on merged design proposal here: kubernetes/community#1680
Issue for tracking PRs: #72828 

**What type of PR is this?**

> /kind feature

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by nolancon** [comment]
I signed it

**Comment by nolancon** [comment]
@lmdaly @ConnorDoyle

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (neolit123) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: does the type has to be public?
if yes, this would need to have an export comment on top.

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (neolit123) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: and instance can be returned directly:
`return Mask{}`

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (neolit123) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: would need an export comment.

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (neolit123) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: export comment.

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (neolit123) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: please change `glog` with `klog` from `k8s.io/klog`.
also ideally `Infof()` calls should be with verbosity level `V(x)`.

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (neolit123) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: could be changed to:
```
return SocketMask(finalMaskValue)...
```

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (neolit123) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: `var maskHolder....` can be moved here.

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (neolit123) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: the extra braces around `len(` are not needed.
needs space after `for j`

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (neolit123) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: errors should start with lowercase.
atoi returns errors that look like this:
```
strconv.Atoi: parsing "some-string": invalid syntax
```
so ideally this message should describe what this conversation is, e.g.:
```
glog.Errorf("could not convert mask character: %v", err)
```

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (neolit123) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: verbosity level should be added unless we want this to always print.

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (neolit123) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: braces around the A & B part are not needed.

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (neolit123) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: could be `return string(byteArr[:])`

**Review by neolit123 — COMMENTED** [review]
thank you for this PR @nolancon 
added some generic comments before running CI tests.

/sig node
/kind feature
/priority important-longterm

**Comment by nolancon** [comment]
@neolit123 comments addressed in last commit.

**Comment by neolit123** [comment]
thanks @nolancon 
please squash the commit fixing comments in the first commit.
up to the maintainers to approve at this point.

**Comment by neolit123** [comment]
/ok-to-test
this will run CI on your PR.

**Comment by spiffxp** [comment]
/hold
ref: https://github.com/kubernetes/enhancements/issues/693#issuecomment-466728227

**Comment by Sreemanti-Ghosh** [comment]
I signed it

**Comment by lmdaly** [comment]
Will be updated based on feedback from the following document: https://docs.google.com/document/d/1UbD1igeBA1sRuqwkQgJ5sJ9sf1rao_LLORWLhwSdGUo/edit?usp=sharing

@klueska

**Comment by derekwaynecarr** [comment]
it looks like https://github.com/kubernetes/kubernetes/pull/72885 prereqs this PR?

**Comment by derekwaynecarr** [comment]
actually, nm, it looks like the previously linked pr should not have a depedency.  i am going to try and work my way through the prs in numeric ordering ;-)

**Review by klueska — COMMENTED** [review]
LGTM pending additional / passing tests.

**Review by klueska — APPROVED** [review]

**Comment by derekwaynecarr** [comment]
the core masking interface and tests look good.

/hold cancel
/approve
/lgtm

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (kad) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: This String() method looks a bit strange, it will print binary representation of mask in opposite order compared to any canonical way of representing masks with lowest bit on the right.

Comment (klueska) [reply pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: Agreed. This should probably be reversed.

Comment (lmdaly) [reply pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: The sockets are set from left to right though - so it should be read that way as a string?

Comment (klueska) [reply pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: The sockets are set as a bit in a uint64, which, if printed as a string, would have bit 0 on the far right and bit 63 on the far left. It's not a huge deal which order these are actually printed in the string representation -- I think @kad's point was just that if we are printing them to look like a binary 64-bit integer, why are we putting the most significant bit on the wrong side.

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (tedyu) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: Should Xor (^) be supported for SocketMask ?

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask.go
Comment (kad) [reply pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: @klueska 
```
return fmt.Sprintf("%064b", *s)
```
will be simple implementation of that function.

Comment (klueska) [reply pkg/kubelet/cm/topologymanager/socketmask/socketmask.go]: This issue is being addressed in https://github.com/kubernetes/kubernetes/pull/78587

**Pull request merged**
